File: mergesort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Merge two sorted halves
void merge(vector<int>& arr, int left, int mid, int right) {
    vector<int> temp;
    int i = left, j = mid + 1;

    while(i <= mid && j <= right) {
        if(arr[i] <= arr[j]) temp.push_back(arr[i++]);
        else temp.push_back(arr[j++]);
    }

    while(i <= mid) temp.push_back(arr[i++]);
    while(j <= right) temp.push_back(arr[j++]);

    for(int k = 0; k < temp.size(); ++k)
        arr[left + k] = temp[k];
}

// Recursive Merge Sort function
void mergeSort(vector<int>& arr, int left, int right) {
    if(left >= right) return;
    int mid = left + (right - left) / 2;

    mergeSort(arr, left, mid);
    mergeSort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}

// Driver code
int main() {
    vector<int> arr = {5, 2, 9, 1, 6, 3};

    mergeSort(arr, 0, arr.size() - 1);

    cout << "Sorted array: ";
    for(int num : arr)
        cout << num << " ";
    cout << endl;

    return 0;
}

File: mergesort.js
function mergeSort(arr) {
    // Base case: array of 0 or 1 elements is already sorted
    if (arr.length <= 1) {
      return arr;
    }
  
    // Split the array into left and right halves
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);
  
    // Recursively sort both halves and merge them
    return merge(mergeSort(left), mergeSort(right));
  }
  
  function merge(left, right) {
    let result = [];
    let i = 0;
    let j = 0;
  
    // Compare elements of left and right arrays and merge them in sorted order
    while (i < left.length && j < right.length) {
      if (left[i] < right[j]) {
        result.push(left[i]);
        i++;
      } else {
        result.push(right[j]);
        j++;
      }
    }
  
    // Add remaining elements (if any)
    return result.concat(left.slice(i)).concat(right.slice(j));
  }
  
  // Example usage:
  let numbers = [38, 27, 43, 3, 9, 82, 10];
  console.log("Sorted array:", mergeSort(numbers));
  